README

CQL-Java - a free CQL compiler, and other CQL tools, for Java


INTRODUCTION
------------

CQL-Java is a Free Software project that provides:

* A set of classes for representing a CQL parse tree (a base CQLNode
  class, CQLBooleanNode and its subclasses, CQLTermNode, etc.)
* A CQLCompiler class (and its lexer) which builds a parse tree given
  a CQL query as input.
* A selection of compiler back-ends to render out the parse tree as:
	* XCQL (the standard XML representation)
	* CQL (i.e. decompiling the parse-tree)
	* PQF (Yaz-style Prefix Query Format)
	* BER code for the Z39.50 Type-1 query
* A random query generator, useful for testing.

CQL is "Common Query Language", a query language designed under
the umbrella of the ZING initiative (Z39.59-International Next
Generation).  The official specification is at
	http://www.loc.gov/standards/sru/cql/
and there's more (and friendlier) information at
	http://zing.z3950.org/cql/index.html

XCQL is "XML CQL", a representation of CQL-equivalent queries in XML
which is supposed to be easier to parse.  The specification is at
	http://docs.oasis-open.org/search-ws/searchRetrieve/v1.0/os/schemas/xcql.xsd
in the form of an XML Schema.

But if you didn't know that, why are you even reading this?  :-)


WHAT'S WHAT IN THIS DISTRIBUTION?
---------------------------------

	README		This file
	Changes		History of releases
	LGPL-2.1	The GNU lesser GPL (see below)
	pom.xml Maven project file to control compilation.
	src     Source-code for the CQL-Java library and tests
	target  The compiled library file, "cql-java.jar" and javadoc
	bin     Simple shell-scripts to invoke CQL programs (parser/lexer/generator)
	util    Various testing and sanity-checking Perl scripts
	etc		Other files: PQF indexes, generator properties, etc.


COMPILATION AND INSTALLATION
----------------------------

The build process is controlled by Maven so compilation is the standard:

	mvn clean install

which generates build artifacts under target/.

"Installation" of this package would consist of putting the bin
directory on your PATH and target/cql-java.jar on your CLASSPATH.


SYNOPSIS
--------

Using the test-harnesses:

	$ CQLParser 'title=foo and author=(bar or baz)'
	$ CQLParser -c 'title=foo and author=(bar or baz)'
	$ CQLParser -p /etc/pqf.properties 'dc.title=foo and dc.author=bar'
	$ CQLLexer 'title=foo and author=(bar or baz)'
		(not very interesting unless you're debugging)
	$ CQLGenerator etc/generate.properties seed 18

Using the library in your own applications:

	import org.z3950.zing.cql.*

	// Building a parse-tree by hand
	CQLNode n1 = new CQLTermNode("dc.author", new CQLRelation("="),
				     "kernighan");
	CQLNode n2 = new CQLTermNode("dc.title", new CQLRelation("all"),
				     "elements style");
	CQLNode root = new CQLAndNode(n1, n2);
	System.out.println(root.toXCQL(0));

	// Parsing a CQL query
	CQLParser parser = new CQLParser();
	CQLNode root = parser.parse("title=dinosaur");
	System.out.print(root.toXCQL(0));
	System.out.println(root.toCQL());
	System.out.println(root.toPQF(config));
	// ... where `config' specifies CQL-qualfier => Z-attr mapping


DESCRIPTION
-----------

See the automatically generated class documentation in the "target"
subdirectory.


AUTHOR
------

Original code and documentation by Mike Taylor, Index Data <mike@indexdata.com>
At present maintained by Jakub Skoczen, Index Data <jakub@indexdata.dk>

	http://www.indexdata.com/cql-java
	http://zing.z3950.org/cql

Please email me with bug-reports, wishlist items, patches, deployment
stories and, of course, large cash donations.


LICENCE
-------

The CQL-Java suite is Free Software, which is pretty much legally
equivalent -- though not morally equivalent -- to Open Source.  See
	http://www.gnu.org/philosophy/free-software-for-freedom.html
for a detailed if somewhat one-sided discussion of the differences,
and particularly of why Free Software is an important idea.

CQL-Java is distributed under version 2.1 of the LGPL (GNU LESSER
GENERAL PUBLIC LICENSE).  A copy of the licence is included in this
distribution, as the file LGPL-2.1.  This licence does not allow you
to restrict the freedom of others to use derived versions of CQL-Java
(i.e. you must share your enhancements), but does let you do pretty
much anything else with it.  In particular, you may deploy CQL-Java as
a part of a non-free larger work.


SEE ALSO
--------

Adam Dickmeiss's CQL compiler, written in C.
Rob Sanderson's CQL compiler, written in Python.
Jakub Skoczen's CQL-js compiler, written in JavaScript https://github.com/indexdata/cql-js
All the other free CQL compilers everyone's going to write  :-)
The "Changes" file, including the "Still to do" section.


Changes

Revision history for the CQL-Java package.
See the bottom of this file for a list of things still to do.

The work for releases 1.0, 1.2 and 1.2.1 was sponsored by the National
Library of Australia, whose help we gratefully acknowledge.

1.13 Wed 10 May 2017 14:44:44 CET
  - add classes for visitor traversal
  - truncation attribute right/left/both/none (thanks to Rustam Usmanov)
  - exact relation == (thanks to Rustam Usmanov)
  - CQLNodeVisitor rewritten to allow for generation of prefix,
  postfix and infix notations
  - enum CQLBoolean

1.12 Mon 17 Mar 2014 16:16:16 CET
  - rewrite the CQL lexer to better conform to CQL spec 
  (e.g retain escapes for wildcards, etc)
  - add jUnit test cases
  - align list of terms (non-compliant) handling with Yaz CQL parser

1.11 Tue  5 Mar 2013 13:38:18 CET
  - Fix problem with parsing unquoted terms that begin with a '.'
  - Fix problem with parsing unquoted terms that contain a '@'
  - allow registration of custom relations
  - optional flag to set strict-but-not-compliant mode that prohibits
  unqouted keywords to be used as terms

1.8  Fri Jul 22 18:02:41 CEST 2011
    - API changes: removed direct access to internal state and introduced 
    getters, removed usages of obsolete collections (Vector, Hashtable),
    simplified method signatues (toXCQL), modified constants to be
    read-only
    - implemented faster XCQL serialization, using pure StringBuilder

1.7  Tue Jul 27 17:35:25 BST 2010
	- Patch from Ralph LeVan to recognise
	  "qualifier.dc.creator"-like properties as well as
	  "index.dc.creator" -- backwards compatible with version
	  prior to 1.0 as well as those subsequent.
	- Reinstate the "exact" relation even when v1.2 is in effect
	  (the default), despite its removal from v1.2 of CQL.  In
	  practice, it seems to be in fairly widespread use, and
	  little real purpose is served in removing it.

1.6  Thu Jun 17 10:56:24 BST 2010
	- Fix the bug where "9x" is lexed as two tokens, a TT_NUMBER
	  followed by a TT_WORD, making unquoted search-terms such as
	  2010-06-17 syntax errors.  Thanks to Erik Hetzner
	  <erik.hetzner@ucop.edu> for the patch.

1.5  Tue May 25 16:44:35 CEST 2010
	- Fix recognition of word-relations to correctly match what is
	  specified in the CQL v1.1 and v1.2 documents.  In
	  particular, "encloses" and "within" are recognised, so that
	  Masterkey's use of "encloses" for IP range authentication
	  works again.
	- Update regression tests accordingly.
	- Extend test/random/ to emit summary line of successful
	  recompilations.

1.4  Thu May 20 00:38:40 BST 2010
	- Add support for interpreting sequences of non-keywords as a
	  single multi-word term rather than a broken INDEX RELATION
	  TERM triplet.  This means that "oxford street" is a valid
	  query rather than a syntax error, and "tottenham court road"
	  is what it looks like rather than a search for the word
	  "court" related by the relation "court" to the index
	  "tottenham".  Note that the word-formed relations "any",
	  "all", "exact" and (for CQL v1.2) "scr" are still treated
	  specially, as are profiled relations in context sets, of the
	  form <set>.<name>.

1.3  Wed May 28 11:02:00 BST 2008
	- Add getSortIndexes() to CQLSortNode, as suggested by Marcel
	  Versteeg <Marcel.Versteeg@KB.nl>

1.2.1  Mon Aug  6 16:54:54 BST 2007
	- Contructors for the CQLParser class are made public (as they
	  should have been all along).

1.2  Tue Jul  3 17:53:28 BST 2007
	- Support for CQL version 1.2 as described at
		http://www.loc.gov/standards/sru/next-version.html
	  The most significant part of the v1.2 change is support for
	  sorting, but as of this writing (29th June) the official SRU
	  maintenance agency web site at  
		http://www.loc.gov/standards/sru/index.html
	  does not include any specification for CQL sorting.
	  Therefore the original proposal at
		http://zing.z3950.org/cql/sorting.html
	  is still considered canonical.

1.0 Fri Jun 29 14:10:28 BST 2007
	- Support for version 1.1 as described at
		http://www.loc.gov/standards/sru/specs/cql.html
	  Since this is a significant leap forward, the version
	  numbering increases to the next major version.
	- "Qualifiers" renamed as "indexes" throughout, to match what
	  is now universal CQL terminology.

	PLEASE NOTE THE FOLLOWING INCOMPATIBLE API CHANGES
	- The class UnknownQualifierException is replaced by
	  UnknownIndexException.
	- The CQLTermNode method getQualifier() is replaced by
	  getIndex().

0.7  Thu Sep  4 22:51:11 2003
	- Support for profiled relations and relation modifiers:
	  recognise any non-key word as a relation or modifier, 
	  rejecting those that are not of the form <prefix>.<name>
	  since these must be explicitly tied to a "context" (what
	  used to be called a qualifier-set or index-set).

0.6  Tue Jul 29 23:33:56 2003
	- Include Ralph's fix for CQLTermNode::toType1BER() to prevent
	  it surrounding multi-word terms in quotes when encoded into
	  BER packet.

0.5  Wed Feb  5 15:50:57 2003
	- Add an extra back-end method,
		byte[] toType1BER(Properties config)
	  which generates an opaque BER-encoded PDU suitable for
	  forwarding to Z39.50 server as a pasrt of a searchRequest
	  APDU.  This was contributed by Ralph Levan <levan@oclc.org>
	- Add srw.resultSet support to the toPFQ() method.
	- Add new resultSetName() method to CQLNode and its
	  subclasses, returning the name of the referenced result-set
	  for a node that is just a result-set reference, and null for
	  anything else.
	- Change regression-test rules for Adam's compiler to use his
	  new front-end cql2xcql instead of the old cqlparse3.
	- Change the names of the anchoring properties in
	  etc/pqf.properties and the CQLTermNode code that examines
	  them, in accordance with Adam's PQF-renderer's behaviour.
	  This gives more flexibility in terms of generating
	  attributes for end-of-field anchoring.  Various other minor
	  tweaks to the pqf.properties file, mostly to the comments.

src/org/z3950/zing/cql/CQLLexer.java
// $Id: CQLLexer.java,v 1.13 2007-06-29 15:38:56 mike Exp $

package org.z3950.zing.cql;
import java.io.StreamTokenizer;
import java.io.StringReader;
import java.util.Hashtable;


// This is a semi-trivial subclass for java.io.StreamTokenizer that:
//	* Has a halfDecentPushBack() method that actually works
//	* Includes a render() method
//	* Knows about the multi-character tokens "<=", ">=" and "<>"
//	* Recognises a set of keywords as tokens in their own right
//	* Includes some primitive debugging-output facilities
// It's used only by CQLParser.
//
class CQLLexer extends StreamTokenizer {
    // New publicly visible token-types
    static int TT_LE        = 1000;	// The "<=" relation
    static int TT_GE        = 1001;	// The ">=" relation
    static int TT_NE        = 1002;	// The "<>" relation
    static int TT_EQEQ      = 1003;	// The "==" relation
    static int TT_AND       = 1004;	// The "and" boolean
    static int TT_OR        = 1005;	// The "or" boolean
    static int TT_NOT       = 1006;	// The "not" boolean
    static int TT_PROX      = 1007;	// The "prox" boolean

    // Support for keywords.  It would be nice to compile this linear
    // list into a Hashtable, but it's hard to store ints as hash
    // values, and next to impossible to use them as hash keys.  So
    // we'll just scan the (very short) list every time we need to do
    // a lookup.
    private class Keyword {
	int token;
	String keyword;
	Keyword(int token, String keyword) {
	    this.token = token;
	    this.keyword = keyword;
	}
    }
    // This should logically be static, but Java won't allow it  :-P
    private Keyword[] keywords = {
	new Keyword(TT_AND, "and"),
	new Keyword(TT_OR,  "or"),
	new Keyword(TT_NOT, "not"),
	new Keyword(TT_PROX, "prox"),
    };

    // For halfDecentPushBack() and the code at the top of nextToken()
    private static int TT_UNDEFINED = -1000;
    private int saved_ttype = TT_UNDEFINED;
    private double saved_nval;
    private String saved_sval;

    // Controls debugging output
    private static boolean DEBUG;

    CQLLexer(String cql, boolean lexdebug) {
	super(new StringReader(cql));
	wordChars('!', '?');	// ASCII-dependency!
	wordChars('[', '`');	// ASCII-dependency!
	quoteChar('"');
	ordinaryChar('=');
	ordinaryChar('<');
	ordinaryChar('>');
	ordinaryChar('/');
	ordinaryChar('(');
	ordinaryChar(')');
	wordChars('\'', '\''); // prevent this from introducing strings
	parseNumbers();
	DEBUG = lexdebug;
    }

    private static void debug(String str) {
	if (DEBUG)
	    System.err.println("LEXDEBUG: " + str);
    }

    // I don't honestly understand why we need this, but the
    // documentation for java.io.StreamTokenizer.pushBack() is pretty
    // vague about its semantics, and it seems to me that they could
    // be summed up as "it doesn't work".  This version has the very
    // clear semantics "pretend I didn't call nextToken() just then".
    //
    private void halfDecentPushBack() {
	saved_ttype = ttype;
	saved_nval = nval;
	saved_sval = sval;
    }

    public int nextToken() throws java.io.IOException {
	if (saved_ttype != TT_UNDEFINED) {
	    ttype = saved_ttype;
	    nval = saved_nval;
	    sval = saved_sval;
	    saved_ttype = TT_UNDEFINED;
	    debug("using saved ttype=" + ttype + ", " +
		  "nval=" + nval + ", sval='" + sval + "'");
	    return ttype;
	}

	underlyingNextToken();
	if (ttype == '<') {
	    debug("token starts with '<' ...");
	    underlyingNextToken();
	    if (ttype == '=') {
		debug("token continues with '=' - it's '<='");
		ttype = TT_LE;
	    } else if (ttype == '>') {
		debug("token continues with '>' - it's '<>'");
		ttype = TT_NE;
	    } else {
		debug("next token is " + render() + " (pushed back)");
		halfDecentPushBack();
		ttype = '<';
		debug("AFTER: ttype is now " + ttype + " - " + render());
	    }
	} else if (ttype == '>') {
	    debug("token starts with '>' ...");
	    underlyingNextToken();
	    if (ttype == '=') {
		debug("token continues with '=' - it's '>='");
		ttype = TT_GE;
	    } else {
		debug("next token is " + render() + " (pushed back)");
		halfDecentPushBack();
		ttype = '>';
		debug("AFTER: ttype is now " + ttype + " - " + render());
	    }
	} else if (ttype == '=') {
	    debug("token starts with '=' ...");
	    underlyingNextToken();
	    if (ttype == '=') {
		debug("token continues with '=' - it's '=='");
		ttype = TT_EQEQ;
	    } else {
		debug("next token is " + render() + " (pushed back)");
		halfDecentPushBack();
		ttype = '=';
		debug("AFTER: ttype is now " + ttype + " - " + render());
	    }
	}

	debug("done nextToken(): ttype=" + ttype + ", " +
	      "nval=" + nval + ", " + "sval='" + sval + "'" +
	      " (" + render() + ")");

	return ttype;
    }

    // It's important to do keyword recognition here at the lowest
    // level, otherwise when one of these words follows "<" or ">"
    // (which can be the beginning of multi-character tokens) it gets
    // pushed back as a string, and its keywordiness is not
    // recognised.
    //
    public int underlyingNextToken() throws java.io.IOException {
	super.nextToken();
	if (ttype == TT_WORD)
	    for (int i = 0; i < keywords.length; i++)
		if (sval.equalsIgnoreCase(keywords[i].keyword))
		    ttype = keywords[i].token;

	return ttype;
    }

    // Simpler interface for the usual case: current token with quoting
    String render() {
	return render(ttype, true);
    }

    String render(int token, boolean quoteChars) {
	if (token == TT_EOF) {
	    return "EOF";
	} else if (token == TT_NUMBER) {
	    if ((double) nval == (int) nval) {
		return new Integer((int) nval).toString();
	    } else {
		return new Double((double) nval).toString();
	    }
	} else if (token == TT_WORD) {
	    return "word: " + sval;
	} else if (token == '"') {
	    return "string: \"" + sval + "\"";
	} else if (token == TT_LE) {
	    return "<=";
	} else if (token == TT_GE) {
	    return ">=";
	} else if (token == TT_NE) {
	    return "<>";
	} else if (token == TT_EQEQ) {
	    return "==";
	}

	// Check whether its associated with one of the keywords
	for (int i = 0; i < keywords.length; i++)
	    if (token == keywords[i].token)
		return keywords[i].keyword;

	// Otherwise it must be a single character, such as '(' or '/'.
	String res = String.valueOf((char) token);
	if (quoteChars) res = "'" + res + "'";
        return res;
    }

    public static void main(String[] args) throws Exception {
	if (args.length > 1) {
	    System.err.println("Usage: CQLLexer [<CQL-query>]");
	    System.err.println("If unspecified, query is read from stdin");
	    System.exit(1);
	}

	String cql;
	if (args.length == 1) {
	    cql = args[0];
	} else {
	    byte[] bytes = new byte[10000];
	    try {
		// Read in the whole of standard input in one go
		int nbytes = System.in.read(bytes);
	    } catch (java.io.IOException ex) {
		System.err.println("Can't read query: " + ex.getMessage());
		System.exit(2);
	    }
	    cql = new String(bytes);
	}

	CQLLexer lexer = new CQLLexer(cql, true);
	int token;
	while ((token = lexer.nextToken()) != TT_EOF) {
	    // Nothing to do: debug() statements render tokens for us
	}
    }
}
